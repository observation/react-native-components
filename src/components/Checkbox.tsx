import React from 'react'
import { View, StyleSheet, TouchableOpacity, ViewStyle, StyleProp } from 'react-native'

import { IconCompleted, IconSelection } from './Icon'
import Log from '../lib/Log'
import theme from '../styles/theme'

type Props = {
  enabled: boolean
  onPress?: () => void
  containerStyle?: StyleProp<ViewStyle>
  iconContainerStyle?: StyleProp<ViewStyle>
  children?: React.ReactNode
  lineHeight?: number
  testID?: string
}

const Checkbox = ({
  enabled,
  onPress,
  containerStyle,
  iconContainerStyle,
  children,
  lineHeight = theme.margin.large,
  testID = 'pressable',
}: Props) => {
  Log.debug('Checkbox')

  return (
    <View style={[styles.containerStyle, containerStyle]}>
      <View style={[styles.iconContainer, iconContainerStyle]}>
        <TouchableOpacity testID={testID} onPress={onPress} activeOpacity={0.5}>
          <View style={[styles.iconInnerContainer, { width: lineHeight, height: lineHeight }]}>
            {enabled ? <IconCompleted /> : <IconSelection />}
          </View>
        </TouchableOpacity>
      </View>
      <View style={{ flexShrink: 1 }}>{children}</View>
    </View>
  )
}

export default Checkbox

const styles = StyleSheet.create({
  containerStyle: {
    flex: 1,
    flexDirection: 'row',
  },
  iconContainer: {
    alignItems: 'flex-start',
  },
  iconInnerContainer: {
    justifyContent: 'center',
  },
})
