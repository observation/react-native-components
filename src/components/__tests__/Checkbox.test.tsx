import React from 'react'
import { Text } from 'react-native'

import { render, fireEvent } from '@testing-library/react-native'

import Checkbox from '../Checkbox'

const mockOnPress = jest.fn()

describe('Checkbox', () => {
  describe('Rendering', () => {
    test('Checked state', () => {
      // WHEN
      const { toJSON } = render(
        <Checkbox containerStyle={{ flex: 1 }} iconContainerStyle={{ flex: 1 }} enabled onPress={mockOnPress}>
          <Text>Test checkbox</Text>
        </Checkbox>,
      )

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('Unchecked state', () => {
      // WHEN
      const { toJSON } = render(
        <Checkbox enabled={false} onPress={mockOnPress}>
          <Text>Test checkbox</Text>
        </Checkbox>,
      )

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })
  })

  test('Handle press', () => {
    // WHEN
    const { getByTestId } = render(
      <Checkbox enabled onPress={mockOnPress}>
        <Text>Test checkbox</Text>
      </Checkbox>,
    )

    // WHEN
    const pressable = getByTestId('pressable')
    fireEvent.press(pressable)

    // THEN
    expect(mockOnPress).toHaveBeenCalled()
  })
})
