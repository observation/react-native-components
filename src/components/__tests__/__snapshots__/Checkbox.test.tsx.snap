// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Rendering Checked state 1`] = `
<View
  style={
    [
      {
        "flex": 1,
        "flexDirection": "row",
      },
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "flex-start",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="pressable"
    >
      <View
        style={
          [
            {
              "justifyContent": "center",
            },
            {
              "height": 24,
              "width": 24,
            },
          ]
        }
      >
        <Icon
          color="#212121"
          name="square-check"
          size={16}
          type="light"
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flexShrink": 1,
      }
    }
  >
    <Text>
      Test checkbox
    </Text>
  </View>
</View>
`;

exports[`Checkbox Rendering Unchecked state 1`] = `
<View
  style={
    [
      {
        "flex": 1,
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "flex-start",
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="pressable"
    >
      <View
        style={
          [
            {
              "justifyContent": "center",
            },
            {
              "height": 24,
              "width": 24,
            },
          ]
        }
      >
        <Icon
          color="#212121"
          name="square"
          size={16}
          type="light"
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flexShrink": 1,
      }
    }
  >
    <Text>
      Test checkbox
    </Text>
  </View>
</View>
`;
